File: RenetikObjc/Library/Classes/Core/Categories/UIKit/UITableView+CSExtension.m
//
//  Created by Rene Dohan on 6/11/12.
//

#import "CSLang.h"
#import "UIView+CSExtension.h"
#import "NSObject+CSExtension.h"
#import "UIViewController+CSExtension.h"
#import "UINavigationController+CSExtension.h"
#import "UIView+CSDimension.h"

#import "UIView+CSLayout.h"

@implementation UITableView (CSExtension)

- (instancetype)construct {
    super.construct;
    self.backgroundColor = UIColor.clearColor;
    return self;
}

- (instancetype)setupTable :(id <UITableViewDelegate, UITableViewDataSource>)parent {
    self.delegate = parent;
    self.dataSource = parent;
    [self reloadData];
    return self;
}

- (instancetype)setupTable :(id <UITableViewDelegate>)delegate :(id <UITableViewDataSource>)dataSource {
    self.delegate = delegate;
    self.dataSource = dataSource;
    [self reloadData];
    return self;
}

- (UIView *)setHeader :(UIView *)view {
    self.tableHeaderView = view;
    return view;
}

- (UIView *)setFooter :(UIView *)view {
    self.tableFooterView = view;
    return view;
}

- (void)deselectSelectedRow {
    [self deselectRowAtIndexPath :self.indexPathForSelectedRow animated :YES];
}

- (void)deselectSelectedRows :(BOOL)animated {
    for (NSIndexPath *path in self.indexPathsForSelectedRows)
        [self deselectRowAtIndexPath :path animated :animated];
}

- (UITableViewCell *)dequeueReusableCell :(NSString *)identifier {
    return [self dequeueReusableCellWithIdentifier :identifier];
}

- (instancetype)hideEmptyCellSplitterBySettingEmptyFooter {
    [self setTableFooterView :[UIView.alloc initWithFrame :CGRectMake(0, 0, 0, 0.1)]];
    return self;
}

- (instancetype)toggleEditingAnimated :(BOOL)animated {
    [self setEditing :!self.isEditing animated :animated];
    return self;
}

- (instancetype)toggleEditing {
    [self setEditing :!self.isEditing animated :YES];
    return self;
}

- (UITableViewCell *)cellView :(Class)viewClass
                              :(void (^)(UITableViewCell *))onCreate {
    var cell = [self dequeueReusableCell :[viewClass className]];
    if (!cell) invokeWith(onCreate, cell = [self createCell :viewClass]);
    return cell;
}

- (UITableViewCell *)createCell :(Class)viewClass {
    let cell = [UITableViewCell.alloc initWithStyle :UITableViewCellStyleDefault reuseIdentifier :[viewClass className]];
    UIView *view = viewClass.construct;
    [cell size :CGSizeMake(self.width, self.rowHeight = view.height)];
    [cell.contentView matchParent];
    [cell.contentView content :view].matchParent;
    return cell;
}



@end

File: Renetik/Renetik/Classes/Core/Extensions/UITableView+CSExtension.swift
//
// Created by Rene on 2018-11-29.
//

import UIKit

@objc public extension UITableView {
    @objc public func reloadKeepSelection() {
        let paths = indexPathsForSelectedRows
        reloadData()
        paths?.forEach { (path: IndexPath?) -> Void in
            selectRow(at: path, animated: false, scrollPosition: .none)
        }
    }

    @nonobjc func cellView<ViewType: UIView>(
        _ cellViewType: ViewType.Type,
        onCreate: @escaping (UITableViewCell, ViewType) -> Void,
        onLoad: @escaping (ViewType) -> Void) -> UITableViewCell {
        var cell = dequeueReusableCell(cellViewType.className())
        if cell.isNil {
            cell = UITableViewCell(style: .default,
                                   reuseIdentifier: cellViewType.className())
            cell!.contentView.matchParent()
            let cellView = cellViewType.init()
            onCreate(cell!, cellView)
            cell!.contentView.content(cellView.construct())
            cell!.width(width, height: cellView.height)
            cellView.matchParent()
        }
        onLoad(cell!.cellView as! ViewType)
        return cell!
    }

    @objc func cell(with identifier: String,
                    style: UITableViewCell.CellStyle,
                    onCreate: ((UITableViewCell) -> Void)? = nil)
        -> UITableViewCell {
        var cell = dequeueReusableCell(identifier)
        if cell.isNil {
            cell = UITableViewCell(style: style, reuseIdentifier: identifier)
            onCreate?(cell!.construct())
        }
        return cell!
    }
}
